File: backend/models/AccountRequest.js
const mongoose = require('mongoose');

const accountRequestSchema = new mongoose.Schema({
  requestType: {
    type: String,
    enum: ['create', 'updateBalance', 'delete'],
    required: true,
  },
  accountId: { type: mongoose.Schema.Types.ObjectId, ref: 'BankAccount', required: true },
  accountName: { type: String, required: true },
  requestedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  requestedByName: { type: String, default: '' },
  previousBalance: { type: Number, default: null },
  newBalance: { type: Number, default: null },
  approvals: {
    developer: { type: Boolean, default: null },
    boss: { type: Boolean, default: null },
    accountant: { type: Boolean, default: null },
    developerUid: { type: String, default: null },
    bossUid: { type: String, default: null },
    accountantUid: { type: String, default: null },
    developerAt: { type: Date, default: null },
    bossAt: { type: Date, default: null },
    accountantAt: { type: Date, default: null },
  },
  status: {
    type: String,
    enum: ['pending', 'partiallyApproved', 'approved', 'rejected'],
    default: 'pending',
  },
  createdAt: { type: Date, default: Date.now },
});

accountRequestSchema.methods.toPublicJSON = function () {
  return {
    id: this._id.toString(),
    requestType: this.requestType,
    accountId: this.accountId.toString(),
    accountName: this.accountName,
    requestedBy: this.requestedBy.toString(),
    requestedByName: this.requestedByName,
    previousBalance: this.previousBalance,
    newBalance: this.newBalance,
    approvals: {
      developer: this.approvals.developer,
      boss: this.approvals.boss,
      accountant: this.approvals.accountant,
      developerUid: this.approvals.developerUid,
      bossUid: this.approvals.bossUid,
      accountantUid: this.approvals.accountantUid,
      developerAt: this.approvals.developerAt ? this.approvals.developerAt.toISOString() : null,
      bossAt: this.approvals.bossAt ? this.approvals.bossAt.toISOString() : null,
      accountantAt: this.approvals.accountantAt ? this.approvals.accountantAt.toISOString() : null,
    },
    status: this.status,
    createdAt: this.createdAt.toISOString(),
  };
};

module.exports = mongoose.model('AccountRequest', accountRequestSchema);

File: lib/models/account_request.dart
import 'payment.dart';

enum AccountRequestType { create, updateBalance, delete }

class AccountRequest {
  final String id;
  final AccountRequestType requestType;
  final String accountId;
  final String accountName;
  final String requestedBy;
  final String requestedByName;
  final double? previousBalance;
  final double? newBalance;
  final PaymentApproval approvals;
  final PaymentStatus status;
  final DateTime createdAt;

  AccountRequest({
    required this.id,
    required this.requestType,
    required this.accountId,
    required this.accountName,
    required this.requestedBy,
    this.requestedByName = '',
    this.previousBalance,
    this.newBalance,
    required this.approvals,
    required this.status,
    required this.createdAt,
  });

  factory AccountRequest.fromMap(Map<dynamic, dynamic> map) {
    final approvals = PaymentApproval.fromMap(
      (map['approvals'] as Map<dynamic, dynamic>?) ?? {},
    );
    return AccountRequest(
      id: map['id'] ?? '',
      requestType: _parseType(map['requestType']),
      accountId: map['accountId'] ?? '',
      accountName: map['accountName'] ?? '',
      requestedBy: map['requestedBy'] ?? '',
      requestedByName: map['requestedByName'] ?? '',
      previousBalance: map['previousBalance'] != null ? (map['previousBalance'] as num).toDouble() : null,
      newBalance: map['newBalance'] != null ? (map['newBalance'] as num).toDouble() : null,
      approvals: approvals,
      status: _computeStatus(approvals),
      createdAt: DateTime.tryParse(map['createdAt']?.toString() ?? '') ?? DateTime.now(),
    );
  }

  static AccountRequestType _parseType(dynamic t) {
    switch (t) {
      case 'updateBalance':
        return AccountRequestType.updateBalance;
      case 'delete':
        return AccountRequestType.delete;
      default:
        return AccountRequestType.create;
    }
  }

  static PaymentStatus _computeStatus(PaymentApproval approvals) {
    // Only developer and boss need to approve account requests (accountant is the initiator)
    if (approvals.developer == false || approvals.boss == false) return PaymentStatus.rejected;
    if (approvals.developer == true && approvals.boss == true) return PaymentStatus.approved;
    if (approvals.developer == true || approvals.boss == true) return PaymentStatus.partiallyApproved;
    return PaymentStatus.pending;
  }

  String get requestTypeDisplayName {
    switch (requestType) {
      case AccountRequestType.create:
        return 'Add Account';
      case AccountRequestType.updateBalance:
        return 'Update Balance';
      case AccountRequestType.delete:
        return 'Remove Account';
    }
  }

  String get statusDisplayName {
    switch (status) {
      case PaymentStatus.pending:
        return 'Pending Approval';
      case PaymentStatus.partiallyApproved:
        return 'Partially Approved';
      case PaymentStatus.approved:
        return 'Approved';
      case PaymentStatus.rejected:
        return 'Rejected';
    }
  }
}
